service/storage_service.cc

```diff
@@ -2497,6 +2497,92 @@ canonical_mutation storage_service::build_mutation_from_join_params(const join_n
     return builder.build();
 }
 
+class join_node_rpc_handshaker : public service::group0_handshaker {
+private:
+    service::storage_service& _ss;
+    const join_node_request_params& _req;
+
+public:
+    join_node_rpc_handshaker(service::storage_service& ss, const join_node_request_params& req)
+            : _ss(ss)
+            , _req(req)
+    {}
+
+    future<bool> perform_handshake(const group0_info& g0_info) override {
+        return _ss.raft_perform_join_handshake(g0_info, _req);
+    }
+};
+
+future<bool> storage_service::raft_perform_join_handshake(const group0_info& g0_info, const join_node_request_params& params) {
+    slogger.info("raft topology: join: sending the join request to {}", g0_info.ip_addr);
+
+    auto result = co_await ser::join_node_rpc_verbs::send_join_node_request(
+            &_messaging.local(), netw::msg_addr(g0_info.ip_addr), g0_info.id, std::move(params));
+    co_return co_await std::visit(overloaded_functor {
+        [this] (const join_node_request_result::ok&) -> future<bool> {
+            // The request has been placed to system.topology.
+            // Now we need to wait until the topology coordinator
+            // either accepts us or rejects us.
+            slogger.info("raft topology: join: request to join placed, waiting  "
+                         "for the response from the topology coordinator");
+
+            _join_node_request_done.set_value();
+
+            // Processing of the response is done in `join_node_response_handler`.
+            // Wait for it to complete.
+            // TODO: Timeouts?
+            co_await _join_node_result.get_shared_future(_abort_source);
+
+            slogger.info("raft topology: join: success");
+            co_return true;
+        },
+        [this] (const join_node_request_result::accepted& acc) -> future<bool> {
+            // We were already added to the topology.
+            _join_node_request_done.set_value();
+
```

Comment:
In the current implementation it is needed, because the joining node responds to the topology coordinator before saving the group0 ID - if there is a crash in between, it will redo the handshake but the topology coordinator will not send `join_node_response`.

I suppose the group0 handshaker interface could be mended so that the order is reversed.

I also wanted to get rid of the possibility for the nodes to retry bootstrap, as discussed [here](https://github.com/scylladb/scylladb/pull/15196#discussion_r1322872458), so `accepted` will be removed anyway. However, saving group0 ID first might still be a good idea, otherwise a node could get added by the topology coordinator, crash and then retry bootstrap, which will fail.

Reply:
> In the current implementation it is needed, because the joining node responds to the topology coordinator before saving the group0 ID - if there is a crash in between, it will redo the handshake but the topology coordinator will not send join_node_response.

So we send `ok` if there is no entry in the topology or the entry is in the request state. But what if an entry is in bootstrapping state but  `join_node_response` was still not issued? As far as I see it may happen since the coordinator first move the node to `bootstrapping` state, then adds it the config and only then issues the response call.

> I also wanted to get rid of the possibility for the nodes to retry bootstrap

OK. I think it should be done before merging then.

Reply:
Now the procedure is not retryable, and we store the group 0 ID before ending the ` join_node_response` RPC handler.